### PR TITLE
feat: authenticated screens and pid receiving

### PR DESCRIPTION
**Pull request by TimoGlastra**
This PR adds support for receiving an SD-JWT VC from the Funke PID issuer, as well as adds the authenticated screens structure (wallet, onboarding, authentication).

It won't work until the following PRs have been released and merged:
- https://github.com/Sphereon-Opensource/OID4VC/pull/131
- https://github.com/Sphereon-Opensource/OID4VC/pull/135
- https://github.com/openwallet-foundation/credo-ts/pull/1983
- https://github.com/openwallet-foundation/credo-ts/pull/1966

### CODE REVIEW

**Comment by TimoGlastra** [comment]
@berendsliedrecht still need to wait for the other PRs for it to work, but the code is complete.

Please note that this is really an intermediate state as I will still be doing a lot of refactoring. So this is just to get the base working flow in. So feel free to comment on whatever, but I may push some things to upcoming PRs if that's easier

File: apps/funke/agent/initialize.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { initializeFunkeAgent } from '@package/agent'
+import { trustedCertificates } from '../constants'
+
+export function initializeAppAgent({ walletKey }: { walletKey: string }) {
+  return initializeFunkeAgent({
+    keyDerivation: 'raw',
+    walletId: 'funke-wallet',
+    walletKey: walletKey,
```
Comment (berendsliedrecht) [inline apps/funke/agent/initialize.ts:8]:
```suggestion
    walletKey,
```

File: apps/funke/agent/initialize.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    walletLabel: 'Funke Wallet',
+    trustedX509Certificates: trustedCertificates,
```
Comment (berendsliedrecht) [inline apps/funke/agent/initialize.ts:10]:
```suggestion
    trustedX509Certificates,
```

File: apps/funke/app.config.js [hunk 3]
```diff
@@ -6,23 +6,14 @@ const variants = {
   development: {
     bundle: '.dev',
     name: ' (Dev)',
-    trustedCertificates: [
```
Comment (berendsliedrecht) [inline apps/funke/app.config.js:9]: Ill likely encounter it later, but why was this removed?

File: apps/funke/app/(app)/_layout.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import { Text } from 'react-native'
+
+import { useSecureUnlock } from '@/agent'
+import { AgentProvider } from '@package/agent'
+
+export default function AppLayout() {
+  const secureUnlock = useSecureUnlock()
+
+  // Wallet is not configured yet. Redirect to onboarding
+  if (secureUnlock.state === 'not-configured') {
```
Comment (berendsliedrecht) [inline apps/funke/app/(app)/_layout.tsx:11]: enum would be nicer, can be done in the future as in improvement.

Comment (berendsliedrecht) [reply apps/funke/app/(app)/_layout.tsx:11]: same as with the href routing.

File: apps/funke/app/(app)/_layout.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  if (secureUnlock.state === 'locked') {
+    return <Redirect href="/authenticate" />
+  }
+
+  // This should show the splash screen
+  if (secureUnlock.state === 'initializing') {
+    return <Text>Loading... initializing</Text>
+  }
+
+  if (!secureUnlock.context.agent) {
```
Comment (berendsliedrecht) [inline apps/funke/app/(app)/_layout.tsx:25]: Maybe just combine it with the initializing state?

File: apps/funke/app/(app)/index.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+export default function Screen() {
+  const { top } = useSafeAreaInsets()
+  // FIXME: should be useReceivePidUseCase as the state is now not updated....
+  const [receivePidUseCase, setReceivePidUseCase] = useState<ReceivePidUseCase>()
+  const [state, setState] = useState<ReceivePidUseCaseState | 'not-initialized'>('not-initialized')
+  const [credential, setCredential] = useState<string>()
+  const { agent } = useAppAgent()
+
+  useEffect(() => {
+    const { remove } = addMessageListener((message) => {
```
Comment (berendsliedrecht) [inline apps/funke/app/(app)/index.tsx:18]: I think the ausweis sdk should have a more explicit name for this function or we should alias it here. Too vague imo.

File: apps/funke/app/(app)/index.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  useEffect(() => {
+    ReceivePidUseCase.initialize({ agent, onStateChange: setState }).then((pidUseCase) => {
+      setReceivePidUseCase(pidUseCase)
+    })
+  }, [agent])
+
+  const nextStep = async () => {
+    if (!receivePidUseCase) return
+    if (state === 'error') return
+
```
Comment (berendsliedrecht) [inline apps/funke/app/(app)/index.tsx:34]:
```suggestion
```

File: apps/funke/app/(app)/index.tsx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    if (state === 'acquire-access-token') return
```
Comment (berendsliedrecht) [inline apps/funke/app/(app)/index.tsx:35]: states should also be an enum.

File: apps/funke/app/(app)/index.tsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    if (state === 'id-card-auth') {
+      console.log('authenticate using id card')
```
Comment (berendsliedrecht) [inline apps/funke/app/(app)/index.tsx:37]: console.log

File: apps/funke/app/authenticate.tsx [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  // TODO: where to put this?
+  void SplashScreen.hideAsync()
+
+  const unlock = async (method: SecureUnlockMethod) => {
+    if (method === 'biometrics') {
+      await secureUnlock.tryUnlockingUsingBiometrics()
+    } else {
+      await secureUnlock.unlockUsingPin('123456')
```
Comment (berendsliedrecht) [inline apps/funke/app/authenticate.tsx:44]: Note to change this ;).

File: apps/funke/app/authenticate.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    }
+  }
+
+  return (
+    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
+      {secureUnlock.canTryUnlockingUsingBiometrics && (
+        <Text onPress={() => unlock('biometrics')}>Unlock using biometrics</Text>
```
Comment (berendsliedrecht) [inline apps/funke/app/authenticate.tsx:51]: did we want to setup translation or skip it for now? I wouldn't change it, but I remember it was a topic a couple of weeks ago.

File: apps/funke/app/onboarding/index.tsx [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import { Redirect } from 'expo-router'
+import { Text, View } from 'react-native'
+
+import { initializeAppAgent, useSecureUnlock } from '@/agent'
```
Comment (berendsliedrecht) [inline apps/funke/app/onboarding/index.tsx:4]: sometimes its @package and sometimes its @/ is there a reason for this?

File: apps/funke/app/onboarding/index.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+  if (secureUnlock.state !== 'not-configured') {
+    return <Redirect href="/" />
+  }
+
+  // TODO: where to put this?
+  void SplashScreen.hideAsync()
+
+  const onboarding = () => {
+    secureUnlock.setup('123456', {
```
Comment (berendsliedrecht) [inline apps/funke/app/onboarding/index.tsx:40]: Same here re change to user input pin

File: apps/funke/tsconfig.json [hunk 14]
```diff
@@ -1,3 +1,9 @@
 {
-	"extends": "../../tsconfig.json"
+	"extends": "../../tsconfig.json",
+	"compilerOptions": {
+		"rootDir": "./",
+		"paths": {
+			"@/*": ["./apps/funke/*"]
```
Comment (berendsliedrecht) [inline apps/funke/tsconfig.json:6]: I am not really fan of the `@/` style. Can we just do `@app/*`? Makes way more sense to me as `@/` is quite vague.

File: apps/funke/use-cases/ReceivePidUseCase.ts [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+  private idCardAuthFlow: AusweisAuthFlow
+  private accessToken?: OpenId4VciRequestTokenResponse
+
+  private onStateChange?: (newState: ReceivePidUseCaseState) => void
+  private currentState: ReceivePidUseCaseState = 'id-card-auth'
+  public get state() {
+    return this.currentState
+  }
+
+  private static SD_JWT_VC_OFFER =
```
Comment (berendsliedrecht) [inline apps/funke/use-cases/ReceivePidUseCase.ts:33]: Are these static values permament or just for testing?

File: packages/secure-store/kdf.ts [hunk 16]
```diff
@@ -10,12 +10,14 @@ import argon2 from 'react-native-argon2'
...
 const derive = async (pin: string, salt: string): Promise<string> => {
+  // Takes about 1.5 second on iPhone 14 Pro
+  // Need to test on Android/different devices as well
   const { rawHash } = await argon2(pin, salt, {
     hashLength: 32,
     mode: 'argon2id',
     parallelism: 4,
-    iterations: 1,
-    memory: 21,
+    iterations: 8,
```
Comment (berendsliedrecht) [inline packages/secure-store/kdf.ts:19]: Why did the values change? Did we not receive these form a spec?

**Review by berendsliedrecht — APPROVED** [review]
Some comments, but due to the nature of the PR I assume it will be fixed in the future.

File: apps/funke/app.config.js (hunk 3 above)
Comment (TimoGlastra) [reply apps/funke/app.config.js:9]: I just couldn't get it to work. The value was not availalbe in the app. So after spending an hour on it, I just moved it to constants file

File: apps/funke/app/authenticate.tsx (hunk 11 above)
Comment (TimoGlastra) [reply apps/funke/app/authenticate.tsx]: yes we talked about it. I don't think we have the time for that

File: apps/funke/app/onboarding/index.tsx (hunk 12 above)
Comment (TimoGlastra) [reply apps/funke/app/onboarding/index.tsx:4]: `@/` is inside the app. But it's a bit confusing. We can do `@app`

File: apps/funke/use-cases/ReceivePidUseCase.ts (hunk 15 above)
Comment (TimoGlastra) [reply apps/funke/use-cases/ReceivePidUseCase.ts:33]: permanent.

File: packages/secure-store/kdf.ts (hunk 16 above)
Comment (TimoGlastra) [reply packages/secure-store/kdf.ts:19]: react-native-argon2 would crash. Did you test it with the previous values? This now takes around 1 sec on my phone. Wanted to do another check later to make it work also on android / older devices

**Pull request merged**